File: swt/src/std.r/file.f
      INTEGER FILTST,RESULT,GETARG,ARGNO
      INTEGER PATHN0(102)
      INTEGER ZEROL0,EXISTS,TYPE,PERMI0,GOTPA0,I
      INTEGER READA0,WRITA0,DUMPED
      INTEGER ARG(102)
      INTEGER PBITS(6)
      INTEGER AAAAA0(29)
      DATA AAAAA0/170,243,186,160,227,225,238,238,239,244,160,244,229,24
     *3,244,160,227,239,238,228,233,244,233,239,238,243,170,238,0/
      PERMI0=0
      TYPE=0
      EXISTS=1
      ZEROL0=0
      READA0=0
      WRITA0=0
      DUMPED=0
      ARGNO=1
      GOTPA0=0
      PBITS(1)=:2000
      PBITS(2)=:1000
      PBITS(3)=:400
      PBITS(4)=:4
      PBITS(5)=:2
      PBITS(6)=:1
10000 IF((-1.EQ.GETARG(ARGNO,ARG,102)))GOTO 10001
        IF((ARG(1).NE.173))GOTO 10002
          CALL MAPSTR(ARG,2)
          IF((ARG(2).EQ.196))GOTO 10003
          IF((ARG(2).EQ.197))GOTO 10003
          IF((ARG(2).EQ.206))GOTO 10003
          IF((ARG(2).EQ.208))GOTO 10003
          IF((ARG(2).EQ.211))GOTO 10003
          IF((ARG(2).EQ.213))GOTO 10003
          IF((ARG(2).EQ.215))GOTO 10003
          IF((ARG(2).EQ.210))GOTO 10003
          IF((ARG(2).EQ.218))GOTO 10003
            CALL USAGE
10003     IF((ARG(2).NE.196))GOTO 10004
            TYPE=:100001
10004     IF((ARG(2).NE.197))GOTO 10005
            EXISTS=1
10005     IF((ARG(2).NE.208))GOTO 10006
            IF((-1.NE.GETARG(ARGNO+1,ARG,102)))GOTO 10007
              CALL USAGE
10007       ARGNO=ARGNO+1
            I=1
            GOTO 10010
10008       I=I+1
10010       IF((I.GT.6))GOTO 10009
              IF((ARG(I).EQ.173))GOTO 10011
                PERMI0=OR(PERMI0,PBITS(I))
10011       GOTO 10008
10009     CONTINUE
10006     IF((ARG(2).NE.210))GOTO 10012
            READA0=1
10012     IF((ARG(2).NE.211))GOTO 10013
            TYPE=:100000
10013     IF((ARG(2).NE.213))GOTO 10014
            TYPE=:100004
10014     IF((ARG(2).NE.215))GOTO 10015
            WRITA0=1
10015     IF((ARG(2).NE.218))GOTO 10016
            ZEROL0=1
10016     IF((ARG(2).NE.206))GOTO 10017
            IF((ARG(3).EQ.197))GOTO 10018
            IF((ARG(3).EQ.215))GOTO 10018
            IF((ARG(3).EQ.210))GOTO 10018
            IF((ARG(3).EQ.218))GOTO 10018
              CALL USAGE
10018       IF((ARG(3).NE.197))GOTO 10019
              EXISTS=-1
10019       IF((ARG(3).NE.210))GOTO 10020
              READA0=-1
10020       IF((ARG(3).NE.215))GOTO 10021
              WRITA0=-1
10021       IF((ARG(3).NE.218))GOTO 10022
              ZEROL0=-1
10022     CONTINUE
10017     GOTO 10023
10002     CALL SCOPY(ARG,1,PATHN0,1)
          GOTPA0=1
10023   ARGNO=ARGNO+1
      GOTO 10000
10001 IF((GOTPA0.NE.0))GOTO 10024
        CALL USAGE
10024 RESULT=FILTST(PATHN0,ZEROL0,PERMI0,EXISTS,TYPE,READA0,WRITA0,DUMPE
     *D)
      IF((RESULT.NE.1))GOTO 10025
        CALL PRINT(-11,'1*n.')
        GOTO 10026
10025   IF((RESULT.NE.0))GOTO 10027
          CALL PRINT(-11,'0*n.')
          GOTO 10028
10027     IF((RESULT.NE.-3))GOTO 10029
            CALL PRINT(-11,'0*n.')
            CALL PRINT(-15,AAAAA0,PATHN0)
10029   CONTINUE
10028 CONTINUE
10026 CALL SWT
      END
      SUBROUTINE USAGE
      CALL ERROR('Usage: file <pathname> -d -[n]e -p twrtwr -[n]r -s -u 
     *-[n]w -[n]z.')
      RETURN
      END
C ---- Long Name Map ----
C zerolength                     zerol0
C pathname                       pathn0
C gotpath                        gotpa0
C writable                       writa0
C readable                       reada0
C permissions                    permi0

File: swt/MFD/SRC/STD.R/FILE.R
# file --- test for file conditions on specified file

integer filtst, result, getarg, argno
character pathname (MAXLINE)
integer zero_length, exists, type, permissions, got_path, i
integer readable, writable, dumped
character arg (MAXLINE)
integer p_bits (6)

permissions = 0   # default permission list:       NO
type = 0          # default type list:             NO
exists = 1        # default finding if file exists:YES
zero_length = 0   # default length test:           NO
readable = 0      # default readability test:      NO
writable = 0      # default writability test:      NO
dumped = 0        # default dumped test:           NO
argno = 1       # first argument number
got_path = NO   # did caller pass a pathname?

#  set up bit flag array for primos permissions
p_bits(1) = :2000       # owner delete/truncate
p_bits(2) = :1000       # owner write permission
p_bits(3) = :400        # owner read permission
p_bits(4) = :4          # non-owner delete/truncate
p_bits(5) = :2          # non-owner write permission
p_bits(6) = :1          # non-owner read permission

#
#  expecting only one pathname per call, find all args and then process
#
while (EOF ~= getarg (argno, arg, MAXLINE)) {
   if (arg(1) == '-'c) {
      call mapstr (arg, UPPER)
#
# check found arg for one of the known arg types
      if (arg(2) ~= 'D'c &&
          arg(2) ~= 'E'c &&
          arg(2) ~= 'N'c &&
          arg(2) ~= 'P'c &&
          arg(2) ~= 'S'c &&
          arg(2) ~= 'U'c &&
          arg(2) ~= 'W'c &&
          arg(2) ~= 'R'c &&
          arg(2) ~= 'Z'c )
            call usage

      if (arg(2) == 'D'c)
         type = :100001
      if (arg(2) == 'E'c)
         exists = 1
      if (arg(2) == 'P'c) {
         if (EOF == getarg (argno + 1, arg, MAXLINE))
            call usage
         argno = argno + 1
         for (i=1; i<=6; i=i+1)
            if (arg(i) ~= '-'c)
               permissions = or (permissions, p_bits (i))
         }
      if (arg(2) == 'R'c)
         readable = 1
      if (arg(2) == 'S'c)
         type = :100000
      if (arg(2) == 'U'c)
         type = :100004
      if (arg(2) == 'W'c)
         writable = 1
      if (arg(2) == 'Z'c)
         zero_length = 1
      if (arg(2) == 'N'c) {
         if (arg(3) ~= 'E'c &&
             arg(3) ~= 'W'c &&
             arg(3) ~= 'R'c &&
             arg(3) ~= 'Z'c )
               call usage

         if (arg(3) == 'E'c)
            exists = -1
         if (arg(3) == 'R'c)
            readable = -1
         if (arg(3) == 'W'c)
            writable = -1
         if (arg(3) == 'Z'c)
            zero_length = -1
         }
      }
#------ end of minus options ------
#
#------ if not a minus option, assume it was a pathname
   else {
      call scopy (arg, 1, pathname, 1)
      got_path = YES
      }

   argno = argno + 1
   }     # end of while loop for arg processing

if (got_path == NO) {      # no pathname... error!
   call usage
   }

result = filtst (pathname, zero_length, permissions, exists, type,
   readable, writable, dumped)
#
#  filtst returns ERR, YES, NO...
if (result == YES)
   call print (STDOUT, "1*n.")
else if (result == NO)
   call print (STDOUT, "0*n.")
else if (result == ERR) {
   call print (STDOUT, "0*n.")
   call print (ERROUT, "*s: cannot test conditions*n"s, pathname)
   }

stop
end

subroutine usage

call error ("Usage: file <pathname> -d -[n]e -p twrtwr -[n]r -s -u -[n]w -[n]z"p)

return
end
